contrib/TESTING.md

```diff
@@ -0,0 +1,95 @@
+
+If you want to run a specific test file, you can specify it directly. For example, to run the tests in `test_wallet.py`, you would use:
+
+```bash
+pytest tests/test_wallet.py
+```
+
+Similarly, you can run a specific test within a file by appending `::` and the test name. For example:
+
+```bash
+pytest tests/test_wallet.py::test_create_new_coldkey
+```
+
+## Writing Tests
+
+When writing tests for Bittensor, you should aim to cover both the "happy path" (where everything works as expected) and any potential error conditions. Here's a basic structure for a test file:
+
+```python
+import pytest
+import bittensor
+
+def test_some_functionality():
+    # Setup any necessary objects or state.
+    wallet = bittensor.wallet()
+
+    # Call the function you're testing.
+    result = wallet.create_new_coldkey()
+
+    # Assert that the function behaved as expected.
+    assert result is not None
+```
+
+In this example, we're testing the `create_new_coldkey` function of the `wallet` object. We assert that the result is not `None`, which is the expected behavior.
+
+## Mocking
+
+In some cases, you may need to mock certain functions or objects to isolate the functionality you're testing. Bittensor uses the `unittest.mock` library for this. Here's an example:
+
+```python
+from unittest.mock import MagicMock
+
+def test_some_functionality():
+    # Create a mock object.
+    mock_wallet = MagicMock()
+
+    # Set the return value for a specific function.
+    mock_wallet.create_new_coldkey.return_value = "mocked_key"
+
+    # Call the function you're testing.
+    result = mock_wallet.create_new_coldkey()
+
+    # Assert that the function behaved as expected.
+    assert result == "mocked_key"
+```
+
+In this example, we're mocking the `create_new_coldkey` function to return a specific value. This allows us to test how our code behaves when `create_new_coldkey` is called, without actually calling the real function.
+
+## Test Coverage
+
+It's important to ensure that your tests cover as much of your code as possible. You can use the `pytest-cov` plugin to measure your test coverage. To use it, first install it with pip:
+
+```bash
+pip install pytest-cov
+```
+
+Then, you can run your tests with coverage like this:
+
+```bash
+pytest --cov=bittensor
+```
+
+This will output a coverage report showing the percentage of your code that's covered by tests.
+
+Remember, while high test coverage is a good goal, it's also important to write meaningful tests. A test isn't very useful if it doesn't accurately represent the conditions under which your code will run.
+
+## Continuous Integration
+
+Bittensor uses GitHub Actions for continuous integration. This means that every time you push changes to the repository, all tests are automatically run. If any tests fail, you'll be notified so you can fix the issue before merging your changes.
+
+You can view the results of these tests in the "Actions" tab of the Bittensor GitHub repository.
```

Comment:
Are we planning to move out bittensor CI from CircleCI to Github Actions? is GH actions already integrated with this repository?

AFAIK, what is integrated right now for CI is CircleCI (what i see in the master branch and the last thing i know)

Reply:
Good catch!  Fixed this.

Reply:
For the record I don't have a strong opinion on this. CI is not really exclusive - it is possible to have both Actions and an external CI system like CircleCI. If CircleCI is what you guys prefer to continue using, that's more than fine.

GitHub Actions are to a certain extent free for open source repositories (like this one), so it wouldn't have caused any extra cost to use them, and they are better integrated with the GitHub UI, but that's pretty much it.